### PR TITLE
Add Rich image editing capabilities to Gutenberg

### LINKED ISSUE #13748: Allow cropping of images from inside the content editor
Carried over from Trac ticket [41466](https://core.trac.wordpress.org/ticket/41466):

In user testing, several people have requested the ability to crop images in the editor context after they have been placed. The expectation seems to be that this should be possible, and the abstraction of having to go to the edit media panel to do this results in the user having to crop the image multiple times to "get it right".

This suggestion has several significant sticking points:
- The idea of cropping images to "fit" within the context of content makes no sense when considering RWD.
- Users who crop images also expect to be able to uncrop them later or in other contexts when the image is reused.
- Hard cropping images by changing the file is becoming a somewhat archaic practice thanks to CSS masking and cropping.

## Possible solutions:
- Provide tools for theme developers to provide display sizes and shapes for images in the editor through CSS and take full advantage of RICG image scaling (related to Trac ticket [35094](https://core.trac.wordpress.org/ticket/35094)).
- Allow "soft" cropping through CSS (introduces complications with RWD, but better than hard cropping).
- Disallow cropping at all.

### PR DESCRIPTION
**Pull request by ajlende**
## Description

Copied from #20637 
Fixes #13748 
Props to @johngodley and @jasmussen for the heavy lifting.

- Adds new image-editor API endpoints in lib/
- Adds new "block" that extends image capabilities in the image block

## How has this been tested?

- Add image
- Click on image and try different editing tools

## Screenshots 

New toolbar items

image

Zoom crop mode

image

## Types of changes

**New features**: Adds crop, flip, rotate capabilities to images.

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-library/src/rich-image/constants.js
Comment (mtias) [inline packages/block-library/src/rich-image/constants.js]: Let's exclude the filters from this PR.

**Comment by jasmussen** [comment]
Nice work, Alex! Cropping is coming along well.

I'm still seeing the old toolbar:

Screenshot 2020-03-31 at 09 19 53

I guess that's temporary as you're working on the refactor?

Let me know if there's anything I can help with. I'd love to polish the pixels, update the icons, center the spinner, this and that!

**Comment by jasmussen** [comment]
Actually scratch that last screenshot with duplicate toolbars, I had some cache lying around.

**Comment by mtias** [comment]
I feel the "Replace" should come after the link and editing tools. What do you think?

**Comment by ajlende** [comment]
I've taken a look into the licensing problem that travis flagged. [react-easy-crop](https://www.npmjs.com/package/react-easy-crop), the only zoom-crop library that seemed suitable, has [tslib (Apache 2.0 licensed)](https://github.com/microsoft/tslib/issues/47) as a dependency. Unfortunately, removing tslib as a dependency isn't so simple. [rollup-plugin-typescript2](https://github.com/ezolenko/rollup-plugin-typescript2/issues/63), used by [bili](), the build tool for react-easy-crop, forces tslib to be included.

There's a few ways to approach removing tslib from react-easy-crop. I could also revert back to [react-image-crop](https://www.npmjs.com/package/react-image-crop) and see if I could open a PR to add a zoom-crop option there.

@mtias How important is zoom-crop to the initial pass of rich-image? And if it is important, is having a free-form option (as opposed to only fixed aspect ratios) also important?

**Comment by ajlende** [comment]
@jasmussen Thanks for taking a look! With the license issue I mentioned above, the cropping interaction might be changing, but I'd say everything else is fair game at this point. Feel free to commit directly to this branch if you are able, or drop some comments for things I can change.

I've also already started looking into centering the spinner. I had to throw an absolute positioned div over the entire block because I don't have access to just the image from rich-image. That caused it to be centered on everything, including the caption, rather than on just the image. So it looks like it might be more than just a CSS change.

Ideally, I could reuse the spinner that already shows up when you use the image upload. However, I noticed that the cover block, that rich-image also augments, doesn't have a spinner when uploading, so that feature isn't available for reuse there.

My first thought was to add the spinner to the cover block and give both the cover and image blocks an `inProgress` prop to hook into from blocks that extend them. Not sure what folks would think about adding extra props aside from what `BlockListBlock` adds.

**Comment by jasmussen** [comment]
> How important is zoom-crop to the initial pass of rich-image? And if it is important, is also having a free-form option (as opposed to only fixed aspect ratios) also important?

I would say both are important insofar as this is where we would very much like to end up. The zooming interaction is a much better experience for block level images, as they keep the block big and easy to interact with.

In that vein — even if it is a fair bit more work, it may still be worth it to still explore addressing this in a way that doesn't violate the licenses necessary, because any diversion will be something we eventually want to revisit.

If there's a way we can build the free-form crop in a way that doesn't technically detract from the zoom-cropping later on, that could potentially be okay for a V1. But what we want to avoid is building in the wrong direction, and having to throw that all out when we are able to revisit to add the zoom cropping.

The license thing really is a bummer, because the cropping interaction in this branch is really really nice. Is there anything we can do?


> Feel free to commit directly to this branch if you are able, or drop some comments for things I can change.

Thanks, I'll push some tweaks when I can! I'd like to:

- Center or appropriately position the spinner
- Add back some of the `transform` effects we had to the flip tool, which flips it instantly with CSS even while the "processing" is happening in the background. This is a nice visual bit of feedback.
- Update the icons.
- Put the "Replace" at the end, to Matías point.

**Comment by ajlende** [comment]
Microsoft/tslib#96 is open now and would fix the licensing problem. Presumably it will be getting [merged soon](https://github.com/microsoft/tslib/issues/47#issuecomment-614918301), so I'm putting this on hold until then.

> We'll give the PR about 10 business days, and if there are no strong objections, we'll pull the change in and publish a new version of tslib with the less-restrictive license.

**Comment by azaozz** [comment]
This looks great! Did some quick testing/played a bit with it. Unlike @jasmussen I'm not a fan of the zoom-crop thing (but this conversation is probably for another place). At first it looked nice, but after trying to use it for a few images it felt quite restrictive/hard to use. Main grips are that as you zoom, part of the image is hidden, so you cannot see what you're doing. Currently that is even worse as the crop area is stuck to 100% width, so most of the image remains hidden. Wondering how "popular" this type of crop UI is/how many of the popular image editing apps use it. Lately I've been using mostly Gimp and this is not in it.

Things that can be fixed/improved now:
- For the zoom-crop, do not set the crop area to 100% width or height. It works a lot better when there's a bit of "air" around it.
- When saving, the changes should be applied to the largest image available, then all sub-sizes created. Currently changes seem applied to the "large" sub-size and cropped images end up only having "medium" and "thumbnail" sub-sizes, and being really small.
- After testing for about 5 min it ended up with about 40 new image files in the uploads dir, despite that I did undo all changes. Wondering what would be a good way to handle multiple actions all at once, so normal editing operations do not have to "edit" the image and add new files. (The old image editor in WP "stores" the actions in an array (which also allows for easy undo/redo) and generates only a temp preview image. The actual changes are applied to the original image on save).
- I'm not that fond of the naming schema. `crop-21-41-35-20-rotate-180-crop-21-41-35-20-IMG_20190825_104317-scaled-300x225.jpg` doesn't sound particularly good (compared to `IMG_20190825_104317-scaled.jpg`) and will easily exceed the max file name length allowed on some filesystems.

**Comment by DanielRosenwasser** [comment]
Just as an FYI, tslib is now relicensed!

**Comment by talldan** [comment]
> How should we enable and limit for testing as experimental?

There are a few ways to manage this. Probably the most relevant for this would be to use the `GUTENBERG_PHASE` flag:
https://github.com/WordPress/gutenberg/blob/master/docs/designers-developers/developers/feature-flags.md

That would prevent what you're working on from shipping in WordPress core, making it only available in the plugin (and generally its ok to ship more experimental features in the plugin if they're in a working state).

If you need to check that you've implemented the flag correctly you can adjust the value in the package.json to simulate a build against core (I think WordPress Core doesn't have the value defined at all):
https://github.com/WordPress/gutenberg/blob/master/package.json#L18 

The second option, if this is very experimental and needs to be opted in to by users, is to set up an option on the experiments page:
https://github.com/WordPress/gutenberg/blob/master/lib/experiments-page.php

That's accessed from the plugin menu:
Screenshot 2020-05-21 at 10 09 14 am

**Comment by jasmussen** [comment]
This is working really well, nice work Alex! I pushed some CSS to center the spinner.

This PR is BIG, and it works well, and I think you should focus on getting it marked as experimental and then merged. 

However as followup PRs we should do a few things:

1. We should show a snackbar notice when a transform that goes through the server finishes: "Image rotated", "Image flipped", "Image cropped", etc.
2. We should update all the new icons.
3. We should fine-tune the cropping experience.

To expand on 2, I'm referring to these icons:

Screenshot 2020-05-21 at 11 39 06

Screenshot 2020-05-21 at 11 39 10

Screenshot 2020-05-21 at 11 39 14

I will draw new ones for you and we can add them to the icon component. 

Sound good?

**Comment by jasmussen** [comment]
Here are some icons for you. 

Crop: 

Screenshot 2020-05-22 at 09 05 37

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17.5 7v8H19V7c0-1.1-.9-2-2-2H9v1.5h8c.3 0 .5.2.5.5zM7 17.5c-.3 0-.5-.2-.5-.5V1H5v4H1v1.5h4V17c0 1.1.9 2 2 2h10.5v4H19v-4h4v-1.5H7z"/></svg>
```

Rotate left and right:

Screenshot 2020-05-22 at 09 19 55

left:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 4V2.2L9 4.8l3 2.5V5.5c3.6 0 6.5 2.9 6.5 6.5 0 2.9-1.9 5.3-4.5 6.2v.2l-.1-.2c-.4.1-.7.2-1.1.2l.2 1.5c.3 0 .6-.1 1-.2 3.5-.9 6-4 6-7.7 0-4.4-3.6-8-8-8zm-7.9 7l1.5.2c.1-1.2.5-2.3 1.2-3.2l-1.1-.9C4.8 8.2 4.3 9.6 4.1 11zm1.5 1.8l-1.5.2c.1.7.3 1.4.5 2 .3.7.6 1.3 1 1.8l1.2-.8c-.3-.5-.6-1-.8-1.5s-.4-1.1-.4-1.7zm1.5 5.5c1.1.9 2.4 1.4 3.8 1.6l.2-1.5c-1.1-.1-2.2-.5-3.1-1.2l-.9 1.1z"/></svg>
```

right:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M15.1 4.8l-3-2.5V4c-4.4 0-8 3.6-8 8 0 3.7 2.5 6.9 6 7.7.3.1.6.1 1 .2l.2-1.5c-.4 0-.7-.1-1.1-.2l-.1.2v-.2c-2.6-.8-4.5-3.3-4.5-6.2 0-3.6 2.9-6.5 6.5-6.5v1.8l3-2.5zM20 11c-.2-1.4-.7-2.7-1.6-3.8l-1.2.8c.7.9 1.1 2 1.3 3.1L20 11zm-1.5 1.8c-.1.5-.2 1.1-.4 1.6s-.5 1-.8 1.5l1.2.9c.4-.5.8-1.1 1-1.8s.5-1.3.5-2l-1.5-.2zm-5.6 5.6l.2 1.5c1.4-.2 2.7-.7 3.8-1.6l-.9-1.1c-.9.7-2 1.1-3.1 1.2z"/></svg>
```

Flip vertical and horizontal

Screenshot 2020-05-22 at 09 39 41

Vertical:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M2 11.2v1.5h20v-1.5H2zM5.5 6c0-.3.2-.5.5-.5h12c.3 0 .5.2.5.5v3H20V6c0-1.1-.9-2-2-2H6c-1.1 0-2 .9-2 2v3h1.5V6zm2 14h2v-1.5h-2V20zm3.5 0h2v-1.5h-2V20zm7-1.5V20c1.1 0 2-.9 2-2h-1.5c0 .3-.2.5-.5.5zm.5-2H20V15h-1.5v1.5zM5.5 18H4c0 1.1.9 2 2 2v-1.5c-.3 0-.5-.2-.5-.5zm0-3H4v1.5h1.5V15zm9 5h2v-1.5h-2V20z"/></svg>
```

Horizontal:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 6v12c0 1.1.9 2 2 2h3v-1.5H6c-.3 0-.5-.2-.5-.5V6c0-.3.2-.5.5-.5h3V4H6c-1.1 0-2 .9-2 2zm7.2 16h1.5V2h-1.5v20zM15 5.5h1.5V4H15v1.5zm3.5.5H20c0-1.1-.9-2-2-2v1.5c.3 0 .5.2.5.5zm0 10.5H20v-2h-1.5v2zm0-3.5H20v-2h-1.5v2zm-.5 5.5V20c1.1 0 2-.9 2-2h-1.5c0 .3-.2.5-.5.5zM15 20h1.5v-1.5H15V20zm3.5-10.5H20v-2h-1.5v2z"/></svg>
```

Aspect ratio I struggled a little with. I initially had this instinct:

Screenshot 2020-05-22 at 09 41 09

But it isn't really clear enough.

Then I thought, how about showing the number itself right there:

Screenshot 2020-05-22 at 09 42 26

I still think that may be the clearest, 16:9 intrinsically says "aspect ratio", but it's a little illegible. I think there might still be something to this.

But I think we should probably start here:

Screenshot 2020-05-22 at 09 45 38

... and then come back to it. SVG:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18.5 5.5h-13c-1.1 0-2 .9-2 2v9c0 1.1.9 2 2 2h13c1.1 0 2-.9 2-2v-9c0-1.1-.9-2-2-2zm.5 11c0 .3-.2.5-.5.5h-13c-.3 0-.5-.2-.5-.5v-9c0-.3.2-.5.5-.5h13c.3 0 .5.2.5.5v9zM6.5 12H8v-2h2V8.5H6.5V12zm9.5 2h-2v1.5h3.5V12H16v2z"/></svg>
```

As I was looking at this one:

Screenshot 2020-05-22 at 09 48 14

I thought to myself: why do we actually have this? I know it switches between portrait and landscape. But I'm not sure it's that intuitive, and it's an extra button that adds complexity. So we might actually want to include this not as a button, but just as an additional aspect ratio in the dropdown:

Image Tools

We could also move the aspect ratio control into the zoom toolbar and make it an overlay, this makes it feel more contextual:

Image Tools 2

In case the overlay is problematic for small images, we could still keep the aspect ratio control there, but move it down below again:

Image Tools 3

**Comment by mtias** [comment]
Would the slider fit in the toolbar?

**Comment by jasmussen** [comment]
With a little will, it sort of can:

Image Tools

That assumes a min-width of 360px (old phones), and that the "cancel" act happens by deselecting the image. 

We could also replace the "Apply" with a checkmark icon instead, but that seems slightly less clear.

**Comment by jasmussen** [comment]
Here's another version, where the block type indicator becomes the "cancel and go back" action, and the Apply button is an icon.

Image Tools

**Comment by mtias** [comment]
Can it go wider if the image allows?

**Comment by jasmussen** [comment]
> Can it go wider if the image allows?

I _think_ we can set it to 100% width and a max-width, and then the slider would flex to the available space. Not a trick we can do for every toolbar, but quite possibly we can for this one.

**Comment by ajlende** [comment]
Now that I have this marked as experimental (thanks @talldan) and all checks pass, I think we should get this rough draft merged. I'll open issues for things that have been mentioned here but haven't yet been addressed, and then we can iterate in future PRs.

@azaozz, @mtias, and @jasmussen, thank you all for trying it out and giving feedback, I really appreciate it! 🙂

**Comment by ajlende** [comment]
I decided to remove rich-image from the cover block. The CSS that worked on the image didn't work on the cover block—another reason that it might make sense as a block-editor component (#22581).

With that, as long as it's properly limited as experimental, this should be ready to merge.

**Review by jasmussen — APPROVED** [review]
This is what I see:

rich image

It is working remarkably well. Because it is marked as experimental, and because this is a big PR, and because this is very impactful, I am happy to have this merged. On the condition, however, that we follow up this PR as soon as we can, with the remaining issues:

- Update with icons from https://github.com/WordPress/gutenberg/pull/21024#issuecomment-632567310, add them to the icon component
- Add snackbar notice to indicate when a server operation completes
- Remove the "rotate aspect ratio to landscape" button in favor of just 16:9 and 9:16 both in the dropdown
- Redesign the toolbar to match this: https://github.com/WordPress/gutenberg/pull/21024#issuecomment-632647853

**Comment by ellatrix** [comment]
Why was this not done in JS, which would be faster and smoother? Right now, each edit needs a round trip to the server, and that with a image no less. Such a big feature/PR should also have gotten a code review before merge.

File: packages/block-library/src/rich-image/rich-image/index.js
Comment (ellatrix) [inline packages/block-library/src/rich-image/rich-image/index.js:1]: Why is this a separate block?

File: packages/block-library/src/rich-image/rich-image/index.js
Comment (ellatrix) [inline packages/block-library/src/rich-image/rich-image/index.js:56]: It would be nice to create new components with hooks.

Comment (ZebulanStanphill) [reply packages/block-library/src/rich-image/rich-image/index.js:56]: In fact, it is recommended by our [guidelines](https://developer.wordpress.org/block-editor/contributors/develop/coding-guidelines/#documenting-wordpresselement-react-components):

> When possible, all components should be implemented as function components, using hooks for managing component lifecycle and state.

**Comment by ellatrix** [comment]
I also can't get it to work locally, I always get an error "Unable to perform the image modification. Please check your media storage."

**Comment by ellatrix** [comment]
My bad, I was trying to edit an image hosted elsewhere. The error messaging should be clearer here.

I just edited an image a few times, and I see a new image in the media library for _every_ edit I did. This will quickly fill people's server storage space. We should keep the original at most (on the server).

File: packages/block-library/src/rich-image/rich-image/index.js
Comment (ajlende) [reply packages/block-library/src/rich-image/rich-image/index.js:1]: We originally worked on this in [Automattic/block-experiments](https://github.com/Automattic/block-experiments/tree/master/blocks/rich-image) where adding the functionality as a separate block was the only option.

#22581 is open for integrating it better and allowing it's usage for developers making other blocks.

File: packages/block-library/src/rich-image/rich-image/index.js
Comment (ajlende) [reply packages/block-library/src/rich-image/rich-image/index.js:56]: Agreed. And, again, this is mostly carried over from Automattic/block-experiments. It'll get mostly rewritten as part of #22581.

File: packages/block-library/src/rich-image/rich-image/index.js
Comment (ellatrix) [inline packages/block-library/src/rich-image/rich-image/index.js:134]: This is causing the block to rerender when it is selected, which is not a great user experience esp. with an image that needs to load.

File: packages/block-library/src/rich-image/rich-image/index.js
Comment (ellatrix) [inline packages/block-library/src/rich-image/rich-image/index.js:147]: Why the div wrapper? I don't see why image editing tools should affect the structure of a block.

**Comment by ajlende** [comment]
> I see the feature added by default in the Gutenberg plugin, or am I missing something?

Experimental, as in Gutenberg phase 2 (https://github.com/WordPress/gutenberg/pull/21024#issuecomment-631836329) as to not ship with core. @talldan also mentioned adding it to the experiments page, which I can go ahead and switch to instead since this is a bigger feature. 

> I'm not talking about previews in the browser, but rather editing it in JS and sending the edited image to the server on save or autosave or some other trigger.

Which benefits do you think we'll get from editing in the browser instead? Using [`WP_Image_Editor`](https://developer.wordpress.org/reference/classes/wp_image_editor/) still seems more straightforward to me. Either way you're going to need to download the saved image since images are saved on the server. In #22579 (batched edits) I was hoping to only do the edit once, ideally on save if I can hook into that event. And I think the CSS animations for previewing from #22566 would be easier and just as smooth as doing them on a canvas.

File: lib/class-wp-rest-image-editor-controller.php
Comment (TimothyBJacobs) [inline lib/class-wp-rest-image-editor-controller.php:127]: This should use `$request['mediaID']`

File: lib/class-wp-rest-image-editor-controller.php
Comment (TimothyBJacobs) [inline lib/class-wp-rest-image-editor-controller.php:71]: `enum` isn't a type, it should instead be `string`.

File: lib/class-wp-rest-image-editor-controller.php
Comment (TimothyBJacobs) [inline lib/class-wp-rest-image-editor-controller.php:90]: `float` is not a json schema type, it should be `number`.

File: lib/image-editor/class-image-editor-crop.php
Comment (TimothyBJacobs) [inline lib/image-editor/class-image-editor-crop.php:53]: Why are we using `floatval` is the types are documented as `integer`? Which is correct?

File: lib/image-editor/class-image-editor-crop.php
Comment (TimothyBJacobs) [inline lib/image-editor/class-image-editor-crop.php:68]: Are these based off an existing pattern that means these need to be `camelCase`? Core uses `snake_case` in PHP land.

**Comment by TimothyBJacobs** [comment]
Instead of using `WP_REST_Request::get_params()` and operating off of that, the code should just use `$request['parameter']` directly.

The REST API also uses `snake_case` not `camelCase` for parameter names. Those should be updated.

The REST API meets every Thursday at 18:00 UTC in slack at #core-restapi. If you aren't seeing feedback from the REST API team, drop a link to the PR during our open floor.

**Comment by gziolo** [comment]
One note, this PR doesn't add a new block but some of the functionality added here makes the impression that a new experimental block is being registered. Given that this functionality is added using a filter, it should be moved to a different package, `@wordpress/editor` might be a good fit if this is planned to work on all edit pages. `@wordpress/block-library` should contain only folders with core blocks.

**Comment by mtias** [comment]
If we can address the outstanding issues promptly, I don't think we need to add it to experiments. The discussion on how much to do on the client vs server is a bit nuanced. It makes sense to me to do all manipulation client side and _then_ commit the updates as a new image on "save". Stacking new images on the server every time would be taxing on storage space and memory usage for shared hosts.

**Comment by TimothyBJacobs** [comment]
Related ticket: https://core.trac.wordpress.org/ticket/49096

File: packages/block-library/src/rich-image/rich-image/index.js
Comment (ajlende) [reply packages/block-library/src/rich-image/rich-image/index.js:147]: The in-progress spinner needs a container to be centered properly on top of the image

Comment (jasmussen) [reply packages/block-library/src/rich-image/rich-image/index.js:147]: Let's rethink the spinner so we don't have to have the div. Can be a fading in and out image which I believe we use to indicate transience elsewhere. Happy to work with you on this 👍

**Comment by ellatrix** [comment]
Ok, even if you could do previewing in JS or even CSS, why do a round trip with an image to the server at all then, if it can be done on the client side? This is double work. Even more than double because the image is being transferred over the network for each edit. And regardless of that, the preview might not be exactly the same. The algorithm on the server might be different. Things like filters suddenly become more difficult to do, while if we just have to handle the client, it’s fairly easy.

**Comment by ajlende** [comment]
Doing the edits on the server was under the assumption that #22579 would be implemented so the request only happens once on save, so you wouldn't be downloading an image per edit. You'd POST the instructions for the edits and only download the final image.

When editing in JS, since you send the image to the server, aren't you going to want to replace the image in the editor with the URL of the saved image to make sure the save worked properly for the preview? That was my assumption, and that way you're both uploading and downloading the image.

The algorithms for crop, rotate, and flip are very straightforward, so I don't think there would be much of a chance for the algorithms to be different for them.

I was thinking that filters would continue to be done in CSS like they are in [Automattic/block-experiments](https://github.com/Automattic/block-experiments/blob/master/blocks/rich-image/style.scss), but I forgot that CSS filters aren't supported in IE11. It could still be done with SVG filters which are supported across browsers—I started a demo for a duotone filter a while ago in https://github.com/Automattic/block-experiments/pull/88.

Still, it might be nice to save the filtered image as well. In that case filters _would_ have more opportunity for differences since you'd have to replicate the SVG filter on the server.

**Comment by ellatrix** [comment]
I think filters are not done in CSS for the same reasons rotate and flip is not done in CSS on the front end? The image itself should be adjusted because the content might not include the CSS or might be distributed by email, RSS, social networks, etc. Additionally, if you download the image from the website, you'll get the original image. Generally people are not expecting anyone to have access to the original images.

With the client already doing the edits, I don't see why we should add the complexity of doing the same thing on the server, if we can just upload the image on save/publish.

I could also imagine Gutenberg being used with another API of a CMS that doesn't support image editing. Editing in JS means using the common API of creating/uploading an image, while editing on the server means additional endpoints are required.

**Comment by azaozz** [comment]
> With the client already doing the edits, I don't see why we should add the complexity of doing the same thing on the server, if we can just upload the image on save/publish.

It would be ideal if that works :) Editing images in the browser (also before uploading them) is the best option. Been looking at that for quite some time, but seems "we're not there yet", the browsers still aren't particularly good image editors.

Known problems with HTML canvas based editors (afaik):
- Insufficient memory/resources on phones and possibly old/slow computers when attempting to edit larger images (the default base size in WP currently is 2560px). See https://github.com/fengyuanchen/cropperjs#known-issues.
- Not able to edit/preserve EXIF data.
- Image quality and/or compression may not be as good as server based editors like ImageMagick.

Would be great if some js based image editors can be evaluated again, there seem to be few that are more popular.

**Comment by jasmussen** [comment]
I think it might be worth separating the conversation on filters. These were experimental, and might not even be a fit for the core project, regardless of previous implementation. If they are (even just making a photo black and white and adjusting contrast would be delightful), it feels a separate conversation. One which I look forward to having, by the way :)

**Comment by ellatrix** [comment]
Regarding EXIF data on JPG images, the edited image would have access to the original image data on the server, so we could adjust the function to get the EXIF data to get it from the original, if so desired. Not sure if that would be considered an ugly fix. Alternatively we can try to preserve it in JS, I would need to research it.

Regarding the quality, are there any articles about this? I don't really see a problem at first sight.

Yes, that's really annoying for phones and old/slow computers... what is the limit on a server?

**Comment by RickorDD** [comment]
When i add the Imageblock the Fontbar hidden the Text.

Imagestest

Remove this Bar to the Top?

**Comment by joppuyo** [comment]
About editing images in JavaScript:

There are tons of features that depend on native WordPress image editor that are completely sidestepped by implementing image editing in JavaScript using canvas, the first thing that comes to mind is [jpeg_quality](https://developer.wordpress.org/reference/hooks/jpeg_quality/) filter that allows users and plugins to alter the JPEG quality.

WordPress also allows hooking into the image processing so plugins like image optimizers can process the images. It's also possible to completely replace the WP_Image_Editor with another class to customize the image processing, for example using a library like [vips](https://github.com/CreunaFI/vips-image-editor) instead of GD or ImageMagick.

If a canvas is used for image editing instead of predictable quality and scaling settings, the results will vary depending on which browser or OS was used when editing the image. Replacing native server-side WordPress functionality with JavaScript may seem trivial at first but it has a lot of consequences that need to be taken into account.

**Comment by TimothyBJacobs** [comment]
Is there a PR somewhere for addressing the REST API related issues? As well as an overall tracking ticket for image editing?

The REST API team is concerned about the amount of server side REST API code that is needed with this current approach so late in the cycle if we want the feature to land in 5.5. There is already a large amount of complex REST API integrations that are slated for [5.5](https://core.trac.wordpress.org/query?status=accepted&status=assigned&status=new&status=reopened&status=reviewing&component=REST+API&milestone=5.5&or&status=accepted&status=assigned&status=new&status=reopened&status=reviewing&focuses=~rest-api&milestone=5.5&col=id&col=summary&col=status&col=owner&col=type&col=priority&col=version&order=priority).

The only way I can see this landing in 5.5 is by doing editing in JS and adding support for https://core.trac.wordpress.org/ticket/49096.

**Comment by ajlende** [comment]
@TimothyBJacobs I didn't get to your edits because I was going to do a batch editing endpoint instead of all the individual ones in #22959. It's taken longer than expected, so I opened up #23037 for those edits now.

**Comment by koraysels** [comment]
can we add more aspect ratios to the crop feature? and how do we go about doing this?

**Comment by paaljoachim** [comment]
I am adding in this associated issue
**Hover color over image block**
https://github.com/WordPress/gutenberg/issues/33939

**Pull request merged**
